modules/porous_flow/src/materials/PorousFlowSingleComponentFluid.C

```diff
@@ -155,48 +172,85 @@ PorousFlowSingleComponentFluid::initQpStatefulProperties()
                                        _temperature[_qp] + _t_c2k);
 }
 
+template <bool is_ad>
 void
-PorousFlowSingleComponentFluid::computeQpProperties()
+PorousFlowSingleComponentFluidTempl<is_ad>::computeQpProperties()
 {
-  const Real Tk = _temperature[_qp] + _t_c2k;
-
   if (_compute_rho_mu)
   {
-    // Density and viscosity, and derivatives wrt pressure and temperature
-    Real rho, drho_dp, drho_dT, mu, dmu_dp, dmu_dT;
-    _fp.rho_mu_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
-                        Tk,
-                        rho,
-                        drho_dp,
-                        drho_dT,
-                        mu,
-                        dmu_dp,
-                        dmu_dT);
-    (*_density)[_qp] = rho;
-    (*_ddensity_dp)[_qp] = drho_dp * _pressure_to_Pascals;
-    (*_ddensity_dT)[_qp] = drho_dT;
-    (*_viscosity)[_qp] = mu / _pressure_to_Pascals / _time_to_seconds;
-    (*_dviscosity_dp)[_qp] = dmu_dp / _time_to_seconds;
-    (*_dviscosity_dT)[_qp] = dmu_dT / _pressure_to_Pascals / _time_to_seconds;
+    if (is_ad)
+    {
+      GenericReal<is_ad> rho, mu;
+      _fp.rho_mu_from_p_T(_porepressure[_qp][_phase_num] * _pressure_to_Pascals,
+                          _temperature[_qp] + _t_c2k,
+                          rho,
+                          mu);
+
+      (*_density)[_qp] = rho;
+      (*_viscosity)[_qp] = mu / _pressure_to_Pascals / _time_to_seconds;
+    }
+    else
+    {
+      // Density and viscosity, and derivatives wrt pressure and temperature
+      Real rho, drho_dp, drho_dT, mu, dmu_dp, dmu_dT;
+      _fp.rho_mu_from_p_T(MetaPhysicL::raw_value(_porepressure[_qp][_phase_num]) *
+                              _pressure_to_Pascals,
+                          MetaPhysicL::raw_value(_temperature[_qp]) + _t_c2k,
+                          rho,
+                          drho_dp,
+                          drho_dT,
+                          mu,
+                          dmu_dp,
+                          dmu_dT);
+      (*_density)[_qp] = rho;
+      (*_ddensity_dp)[_qp] = drho_dp * _pressure_to_Pascals;
+      (*_ddensity_dT)[_qp] = drho_dT;
+      (*_viscosity)[_qp] = mu / _pressure_to_Pascals / _time_to_seconds;
+      (*_dviscosity_dp)[_qp] = dmu_dp / _time_to_seconds;
+      (*_dviscosity_dT)[_qp] = dmu_dT / _pressure_to_Pascals / _time_to_seconds;
```

Comment:
A general comment I have is that these dy_dx quantities will themselves be functions of the nonlinear degrees of freedom. I have been slowly trying to add more emphasis to the the 5-parameter AD overloads. Just something to maybe think about

Reply:
I'll have to take a look at these - I remember getting a GitHub notification about them but didn't really take a good look